sway-lib-std/src/bytes.sw

```diff
@@ -688,6 +693,10 @@ impl Bytes {
     /// }
     /// ```
     pub fn append(ref mut self, ref mut other: self) {
+        if __addr_of::<Self>(self) == __addr_of::<Self>(other) {
```

Comment:
Pointer equality is not semantic equality. This check is Undefined Behavior.

The compiler is perfectly allowed to split your struct into multiple ones or unify it with another of the same value so this will not have any coherent.

Reply:
This check is to ensure the two `Bytes` structs of `self` and `other` do not point to the same struct and works as intended.

`bytes.ptr() == bytes.ptr()` would not work as when `new()` is called, we alloc size of 0 which simply [returns the top of the heap](https://docs.fuel.network/docs/specs/fuel-vm/instruction-set/#aloc-allocate-memory).

This passes:
```sway
assert(alloc(0) == alloc(0));
```

With this check we do not encounter this problem:

```sway
let bytes_1 = Bytes::new();
let bytes_2 = Bytes::new();

// This fails, they have the same pointer
assert(bytes_1.ptr() != bytes_2.ptr());

let bytes_3 = Bytes::with_capacity(1);
let bytes_4 = Bytes::with_capacity(1);

// This passes
assert(bytes_3.ptr() != bytes_4.ptr());
```

Reply:
You don't understand. Just because this works most of the time does not mean it would work all the time or is even legal Sway.

You can't assume that just because the pointers of two structs are equal, these are the same struct or that these are not different structs. The compiler is allowed to break that assumption.

The code of append with that check would behave differently depending on what particular aliasing optimizations are done or not, and that's illegal.

Reply:
What would happen with

```sway
fn id<T>(v: T) -> T { v }

let mut b= Bytes::with_capacity(1);
// fill b with some data
b.append(id(b));
````

I have the impression that `self` and `other` inside the `append` function will not have the same address, but they will still point to the same buffer.

This happen because `sway` still do not move semantics.

But even before all that, we do need to define the semantics of this method better. We can make this method always work, copying `other`'s buffer into a temporary buffer first.

Wasteful, but will always work. This c++ `std::memmove()` semantics (https://en.cppreference.com/w/cpp/string/byte/memmove)

> The objects may overlap: copying takes place as if the characters were copied to a temporary character array and then the characters were copied from the array to dest.

To optimize this we would need to make sure buffers do not intersect:  would need to check if both "buffers" (from `ptr`  until `ptr + len`) do no intersect;

Avoiding the temp buffer in this case.

> bytes.ptr() == bytes.ptr() would not work as when new() is called, we alloc size of 0 which simply returns the top of the heap.

This could easily be fixed moving the `ptr` checking after the "size optimization", right? If `other.len() == 0` we do not need to do anything.

Reply:
Frankly, I'd rather we sidestep the problem and avoid having to deal with implementing move semantics for now.